logstash-core/spec/logstash/compiler/compiler_spec.rb

```diff
@@ -560,18 +560,92 @@ def compose(*statements)
             end
 
             describe "'=='" do
-              let(:expression) { "[foo] == 5"}
+              let(:lhs) { fail(NotImplementedError) }
+              let(:expression) { "[foo] == #{rhs}"}
 
-              it "should compile correctly" do
-                expect(c_expression).to ir_eql(j.eEq(j.eEventValue("[foo]"), j.eValue(5.to_java)))
+              context 'when RHS represents an integer' do
+                let(:rhs) { '5' }
+                it "should compile correctly" do
+                  expect(c_expression).to ir_eql(j.eEq(j.eEventValue("[foo]"), j.eValue(5.to_java)))
+                end
+              end
+
+              context 'when RHS represents an empty array' do
+                let(:rhs) { '[]' }
+                it "should compile correctly" do
+                  expect(c_expression).to ir_eql(j.eEq(j.eEventValue("[foo]"), j.eValue([])))
+                end
+              end
+
+              # multi-item arrays are not supported because a single-item arrays in this context are
+              # syntactically ambiguous with _selector_s, which take a higher precedence; if/when support
+              # is added for unambiguously declaring a single-item array as a condition clause, this spec
+              # can be removed.
+              context 'when RHS represents a multi-item array' do
```

Comment:
Thinking more about this case, I wonder if we should allow multi-element arrays as rvalues?

As we discussed, there is ambiguity for field references vs single-element arrays, but for multi-element there is no ambiguity. This said, we 6.2 crashes on array comparison anyway: `"foo" == [1,2,3]` hits a bug and throws an exception.

Thoughts on allowing multi-element `[foo] == [1,2,3]` ?

It's ok if multi-element waits until a future PR; I dont' want to stall this bug fix.

Reply:
I think that adding support for multi-element arrays when we don't have a way to unambiguously represent single-element arrays is an opportunity for more confusion and surprises than it's worth, e.g.:

 - `if [foo] == []`: "if the event's `foo` is an empty array"
 - `if [foo] == [1]`: "if the event's `foo` field is equal to the event's `1` field" - wut?
 - `if [foo] == [1,2]`: "if the event's `foo` is equal to the array `[1,2]`"

Adding support for multi-element arrays hints at a continuity that we cannot currently represent, so while we don't currently have a way to reject input like `if [foo] == [1]`, at least we aren't encouraging a syntax where a user's intent to remove one element from an array drastically changes the meaning to do something completely different than their intent.